### PR TITLE
stabilize const_fn_floating_point_arithmetic

### LINKED ISSUE #57241: Tracking issue for `#![feature(const_fn_floating_point_arithmetic)]`
Sub-tracking issue for rust-lang/rfcs#911.

This tracks arithmetic for floating point types in `const fn`. These operations are already stable in `const`/`static`, and are also being promoted on stable (`&(1.2*2.1)` has lifetime `'static`). However, in all of these situations, the code is *definitely* executed at compile-time -- only `const fn` code could be executed *both* at compile-time and run-time, making possible differences in FP behavior observable. Also see https://github.com/rust-lang/rust/issues/77745.

## Open Questions

* https://github.com/rust-lang/rust/issues/77745: Must a `const fn` behave exactly the same at runtime as at compile-time?
* See https://github.com/rust-lang/unsafe-code-guidelines/issues/237 for a collection of issues around floating-point semantics.

### PR DESCRIPTION
**Pull request by RalfJung**
Part of https://github.com/rust-lang/rust/issues/128288
Fixes https://github.com/rust-lang/rust/issues/57241

The existing test `tests/ui/consts/const_let_eq_float.rs`  ([link](https://github.com/RalfJung/rust/blob/const_fn_floating_point_arithmetic/tests/ui/consts/const_let_eq_float.rs)) covers the basics, and also Miri has extensive tests covering the interpreter's float machinery. Also, that machinery can already be used on stable inside `const`/`static` initializers, just not inside `const fn`.

This was explicitly called out in https://github.com/rust-lang/rfcs/pull/3514 so in a sense t-lang just recently already FCP'd this, but let's hear from them whether they want another FCP for the stabilization here or whether that was covered by the FCP for the RFC.
Cc @rust-lang/lang

### Open items

- [x] Update the Reference: https://github.com/rust-lang/reference/pull/1566

### CODE REVIEW

File: tests/ui/internal/internal-unstable-const.rs
Comment (RalfJung) [inline tests/ui/internal/internal-unstable-const.rs:1]: I removed this test since it seems fully covered by the min_const_fn tests.

**Comment by traviscross** [comment]
Thanks @RalfJung.  This looks consistent with what we adopted in RFC 3514.  Let's go ahead with an expectation of a quick and uneventful:

@rfcbot fcp merge

File: tests/ui/consts/const-extern-fn/const-extern-fn-min-const-fn.rs [hunk 1]
```diff
@@ -1,9 +1,5 @@
 #![feature(const_extern_fn)]
 
-const extern "C" fn unsize(x: &[u8; 3]) -> &[u8] { x }
-const unsafe extern "C" fn closure() -> fn() { || {} }
-const unsafe extern "C" fn use_float() { 1.0 + 1.0; }
```
Comment (nnethercote) [inline tests/ui/consts/const-extern-fn/const-extern-fn-min-const-fn.rs:5]: Why are `unsize` and `closure` removed?

Comment (nnethercote) [reply tests/ui/consts/const-extern-fn/const-extern-fn-min-const-fn.rs:5]: Oh, they are moved into a different test below.

**Comment by nnethercote** [comment]
Looks good, r=me once the FCP is satisfied.

**Comment by ehuss** [comment]
Before this is merged, can we please have a reference documentation PR posted and reviewed?

**Comment by RalfJung** [comment]
:tada: 

@bors r=nnethercote

**Pull request merged**